### PR TITLE
Add indented_writer_base<T> to library

### PR DESCRIPTION
**Pull request by jonitis**
Start using common code for tool/python and samples/meta_reader.
This just takes indentation code from PR https://github.com/Microsoft/xlang/pull/260.
The benefit to python tool is that it does not output indented empty lines, since new code explicitly checks for that.

Changes in src/tool/python/main.cpp are whitespace-only to remove trailing whitespaces.

### CODE REVIEW

**Comment by devhawk** [comment]
FYI, I'm working towards the first release of the python projection. I love the idea of pulling the indented code writer out of python and into the shared library, but I need to put off merging this into master for a couple weeks.

**Comment by kennykerr** [comment]
Hey @jonitis, Harry has left the team. Are you still interested in contributing this update? Note that we don't have anyone actively working on Python at the moment so it may require some effort to get this up and running, but I'm happy to help as time allows.

**Comment by jonitis** [comment]
Hello, Kenny. I've made this patch after suggestion of Harry Pierson, since both python_tool and meta_reader (former idl tool) now use identical approach for indenting text. It made sense to move common functionality into library, so that both tools do not have to duplicate code. The python tool will benefit from small change that I've already made in meta_reader - there is no indentation for empty lines.
Feel free to decide whether you want to accept this patch. If you are interested, I can refresh the patch (in two weeks) on top of latest head to resolve the conflicts.

File: src/library/text_writer.h [hunk 1]
```diff
@@ -290,6 +290,102 @@ namespace xlang::text
...
+                _writer.indent += _offset;
+            }
+
+            ~indent_guard()
+            {
+                _writer.indent -= _offset;
+            }
+
+        private:
+            indented_writer_base<T>& _writer;
```
Comment (kennykerr) [inline src/library/text_writer.h]: We use the `m_xxx` naming convention for members.

File: src/library/text_writer.h [hunk 2]
```diff
@@ -290,6 +290,102 @@ namespace xlang::text
...
+            indent = 0;
+
+            auto result = writer_base<T>::write_temp(value, args...);
+
+            indent = restore_indent;
+
+            return result;
+        }
+
+        int32_t indent{};
```
Comment (kennykerr) [inline src/library/text_writer.h]: m_indent

File: src/library/text_writer.h [hunk 3]
```diff
@@ -290,6 +290,102 @@ namespace xlang::text
         std::vector<char> m_first;
     };
 
+
+    template <typename T>
+    struct indented_writer_base : public writer_base<T>
```
Comment (kennykerr) [inline src/library/text_writer.h]: The `public` is redundant.

File: src/library/text_writer.h [hunk 4]
```diff
@@ -290,6 +290,102 @@ namespace xlang::text
...
+    {
+        struct indent_guard
+        {
+            explicit indent_guard(indented_writer_base<T>& w, int32_t offset = 1) noexcept : _writer(w), _offset(offset)
+            {
+                _writer.indent += _offset;
+            }
+
+            ~indent_guard()
```
Comment (kennykerr) [inline src/library/text_writer.h]: noexcept

File: src/library/text_writer.h [hunk 5]
```diff
@@ -290,6 +290,102 @@ namespace xlang::text
...
+            {
+                _writer.indent -= _offset;
+            }
+
+        private:
+            indented_writer_base<T>& _writer;
+            int32_t _offset;
```
Comment (kennykerr) [inline src/library/text_writer.h]: Initialize just for the sake of paranoia.

File: src/library/text_writer.h (hunk 1 above)
Comment (jonitis) [reply src/library/text_writer.h]: Fixed. It was copy-pasted from pywinrt tool which had slightly different style.

File: src/library/text_writer.h (hunk 2 above)
Comment (jonitis) [reply src/library/text_writer.h]: fixed

File: src/library/text_writer.h (hunk 3 above)
Comment (jonitis) [reply src/library/text_writer.h]: fixed

File: src/library/text_writer.h (hunk 4 above)
Comment (jonitis) [reply src/library/text_writer.h]: fixed

File: src/library/text_writer.h (hunk 5 above)
Comment (jonitis) [reply src/library/text_writer.h]: Sure :)

**Pull request merged**